sycl/source/detail/context_impl.cpp

```diff
@@ -47,8 +47,24 @@ context_impl::context_impl(const std::vector<sycl::device> Devices,
       MSupportBufferLocationByDevices(NotChecked) {
   MPlatform = detail::getSyclObjImpl(MDevices[0].get_platform());
   std::vector<sycl::detail::pi::PiDevice> DeviceIds;
-  for (const auto &D : MDevices) {
-    DeviceIds.push_back(getSyclObjImpl(D)->getHandleRef());
+  for (size_t I = 0, E = MDevices.size(); I != E; ++I) {
+    if (MDevices[I].has(aspect::ext_oneapi_is_composite)) {
+      // Component devices are considered to be descendent devices from a
+      // composite device and therefore context created for a composite
+      // device should also work for a component device.
+      // In order to achieve that, we implicitly add all component devices to
+      // the list if a composite device was passed by user to us.
+      std::vector<device> ComponentDevices =
+          MDevices[I]
+              .get_info<
+                  ext::oneapi::experimental::info::device::component_devices>();
+      for (const auto &CD : ComponentDevices) {
+        DeviceIds.push_back(getSyclObjImpl(CD)->getHandleRef());
+        MDevices.push_back(CD);
```

Comment:
Actually, this is a very good question. I honestly don't have a clear answer. You could argue that component devices, since they are part of the composite device, they should indeed be returned by `context::get_devices()`, particularly considering that you can use them to run kernels on them. Oppositely, you could argue that only devices that explicitly added to the context should be returned.

Reply:
No, we don't want `context::get_devices` to return any "extra" devices.  It should return only those devices that were passed to the constructor.

We have the same situation with sub-devices.  Consider a context _C_ that is created from a single root device _D_, and consider the case when the application creates two sub-devices _SD1_ and _SD2_ from _D_.  The sub-devices _SD1_ and _SD2_ are descendent devices of _D_, which means that you can submit a kernel to either _SD1_ or _SD2_ on a queue using context _C_ even though that context doesn't contain either _SD1_ or _SD2_.  Calling `C.get_devices` returns only _D_ (not _SD1_ or _SD2_).

We implemented this by pushing the responsibility for "descendent devices" down to the Level Zero and OpenCL drivers.  In my example above, the Level Zero context _C_ contains only _D_.  The driver still allows kernels to be submitted to sub-devices _SD1_ or _SD2_ even though these sub-devices aren't in context _C_.

I think we should do the same for composite devices.  When Level Zero and OpenCL see a composite device in a context, they should treat all component devices as "descendent devices".

Reply:
Updated the patch to avoid exposing implicitly added devices to user in d5a4ef3b9c699e47dd03e7bc24ed0bdc32ae3d3b

I will speak with L0 folks to see if we can push this logic to them